packages/client/src/grpc-retry.ts

```diff
@@ -1,37 +1,85 @@
-import {
-  InterceptingCall,
-  Interceptor,
-  ListenerBuilder,
-  Metadata,
-  RequesterBuilder,
-  StatusObject,
-} from '@grpc/grpc-js';
+import { InterceptingCall, Interceptor, ListenerBuilder, RequesterBuilder, StatusObject } from '@grpc/grpc-js';
 import * as grpc from '@grpc/grpc-js';
 
 export interface GrpcRetryOptions {
-  /** Maximum number of allowed retries. Defaults to 10. */
-  maxRetries: number;
-
   /**
-   * A function which accepts the current retry attempt (starts at 0) and returns the millisecond
+   * A function which accepts the current retry attempt (starts at 1) and returns the millisecond
    * delay that should be applied before the next retry.
    */
-  delayFunction: (attempt: number) => number;
+  delayFunction: (attempt: number, status: StatusObject) => number;
 
   /**
    * A function which accepts a failed status object and returns true if the call should be retried
    */
-  retryableDecider: (status: StatusObject) => boolean;
+  retryableDecider: (attempt: number, status: StatusObject) => boolean;
+}
+
+/**
+ * Options for the backoff formula: `factor ^ attempt * initialIntervalMs(status) * jitter(maxJitter)`
+ */
+export interface BackoffOptions {
+  /**
+   * Exponential backoff factor
+   *
+   * @default 2
+   */
+  factor: number;
+
+  /**
+   * Maximum number of attempts
+   *
+   * @default 10
+   */
+  maxAttempts: number;
+  /**
+   * Maximum amount of jitter to apply
+   *
+   * @default 0.1
+   */
+  maxJitter: number;
+  /**
+   * Function that returns the "initial" backoff interval based on the returned status.
+   *
+   * The default is 1 second for RESOURCE_EXHAUSTED errors and 20 millis for other retryable errors.
+   */
+  initialIntervalMs(status: StatusObject): number;
 }
 
-export function defaultGrpcRetryOptions(): GrpcRetryOptions {
+/**
+ * Add defaults as documented in {@link BackoffOptions}
+ */
+function withDefaultBackoffOptions({
+  maxAttempts,
+  factor,
+  maxJitter,
+  initialIntervalMs,
+}: Partial<BackoffOptions>): BackoffOptions {
   return {
-    maxRetries: 10,
-    delayFunction: backOffAmount,
-    retryableDecider: isRetryableError,
+    maxAttempts: maxAttempts ?? 10,
+    factor: factor ?? 2,
+    maxJitter: maxJitter ?? 0.1,
+    initialIntervalMs: initialIntervalMs ?? defaultInitialIntervalMs,
   };
 }
 
+/**
+ * Generates the default retry behavior based on given backoff options
+ */
+export function defaultGrpcRetryOptions(options: Partial<BackoffOptions> = {}): GrpcRetryOptions {
+  const { maxAttempts, factor, maxJitter, initialIntervalMs } = withDefaultBackoffOptions(options);
+  return {
+    delayFunction(attempt, status) {
+      return factor ** attempt * initialIntervalMs(status) * jitter(maxJitter);
```

Comment:
I wish it were this easy in Go. Neither in the gRPC interceptor nor in the poll throttler are you allowed to differentiate error types. This is all hidden, so we have to copy/paste a bunch of third party code or have a bad algorithm. Not sure which will be chosen yet.